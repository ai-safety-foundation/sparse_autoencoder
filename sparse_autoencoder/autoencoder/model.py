"""The Sparse Autoencoder Model."""

from typing import final

import torch
from torch.nn.parameter import Parameter

from sparse_autoencoder.autoencoder.abstract_autoencoder import AbstractAutoencoder
from sparse_autoencoder.autoencoder.components.linear_encoder import LinearEncoder
from sparse_autoencoder.autoencoder.components.tied_bias import TiedBias, TiedBiasPosition
from sparse_autoencoder.autoencoder.components.unit_norm_decoder import UnitNormDecoder
from sparse_autoencoder.tensor_types import (
    InputOutputActivationBatch,
    InputOutputActivationVector,
    LearnedActivationBatch,
)


@final
class SparseAutoencoder(AbstractAutoencoder):
    """Sparse Autoencoder Model."""

    geometric_median_dataset: InputOutputActivationVector
    """Estimated Geometric Median of the Dataset.

    Used for initialising :attr:`tied_bias`.
    """

    tied_bias: InputOutputActivationBatch
    """Tied Bias Parameter.

    The same bias is used pre-encoder and post-decoder.
    """

    n_input_features: int
    """Number of Input Features."""

    n_learned_features: int
    """Number of Learned Features."""

    _pre_encoder_bias: TiedBias

    _encoder: LinearEncoder

    _decoder: UnitNormDecoder

    _post_decoder_bias: TiedBias

    @property
    def pre_encoder_bias(self) -> TiedBias:
        """Pre-encoder bias."""
        return self._pre_encoder_bias

    @property
    def encoder(self) -> LinearEncoder:
        """Encoder."""
        return self._encoder

    @property
    def decoder(self) -> UnitNormDecoder:
        """Decoder."""
        return self._decoder

    @property
    def post_decoder_bias(self) -> TiedBias:
        """Post-decoder bias."""
        return self._post_decoder_bias

    def __init__(
        self,
        n_input_features: int,
        n_learned_features: int,
        geometric_median_dataset: InputOutputActivationVector | None = None,
    ) -> None:
        """Initialize the Sparse Autoencoder Model.

        Args:
            n_input_features: Number of input features (e.g. `d_mlp` if training on MLP activations
                from TransformerLens).
            n_learned_features: Number of learned features. The initial paper experimented with 1 to
                256 times the number of input features, and primarily used a multiple of 8.
            geometric_median_dataset: Estimated geometric median of the dataset.
        """
        super().__init__()

        self.n_input_features = n_input_features
        self.n_learned_features = n_learned_features

        # Store the geometric median of the dataset (so that we can reset parameters). This is not a
        # parameter itself (the tied bias parameter is used for that), so gradients are disabled.
        if geometric_median_dataset is not None:
            self.geometric_median_dataset = geometric_median_dataset.clone()
            self.geometric_median_dataset.requires_grad = False
        else:
            self.geometric_median_dataset = torch.zeros(n_input_features)
            self.geometric_median_dataset.requires_grad = False

        # Initialize the tied bias
        self.tied_bias = Parameter(torch.empty(n_input_features))
        self.initialize_tied_parameters()

        self._pre_encoder_bias = TiedBias(self.tied_bias, TiedBiasPosition.PRE_ENCODER)

        self._encoder = LinearEncoder(
            input_features=n_input_features, learnt_features=n_learned_features
        )

        self._decoder = UnitNormDecoder(
            learnt_features=n_learned_features, decoded_features=n_input_features
        )

        self._post_decoder_bias = TiedBias(self.tied_bias, TiedBiasPosition.POST_DECODER)

    def forward(
        self,
        x: InputOutputActivationBatch,
    ) -> tuple[
        LearnedActivationBatch,
        InputOutputActivationBatch,
    ]:
        """Forward Pass.

        Args:
            x: Input activations (e.g. activations from an MLP layer in a transformer model).

        Returns:
            Tuple of learned activations and decoded activations.
        """
        x = self._pre_encoder_bias(x)
        learned_activations = self._encoder(x)
        x = self._decoder(learned_activations)
        decoded_activations = self._post_decoder_bias(x)
        return learned_activations, decoded_activations

    def initialize_tied_parameters(self) -> None:
        """Initialize the tied parameters."""
        # The tied bias is initialised as the geometric median of the dataset
        self.tied_bias.data = self.geometric_median_dataset

    def reset_parameters(self) -> None:
        """Reset the parameters."""
        self.initialize_tied_parameters()
        for module in self.network:
            if "reset_parameters" in dir(module):
                module.reset_parameters()
