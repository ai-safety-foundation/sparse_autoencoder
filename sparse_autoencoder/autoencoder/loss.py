"""Loss function for the Sparse Autoencoder."""
from jaxtyping import Float
import torch
from torch import Tensor
from torch.nn.functional import mse_loss


def reconstruction_loss(
    input_activations: Float[Tensor, "*batch input_activations"],
    output_activations: Float[Tensor, "*batch input_activations"],
) -> Tensor:
    """Reconstruction Loss (MSE).

    MSE reconstruction loss is calculated as the mean squared error between each each input vector
    and it's corresponding decoded vector. The original paper found that models trained with some
    loss functions such as cross-entropy loss generally prefer to represent features
    polysemantically, whereas models trained with MSE may achieve the same loss for both
    polysemantic and monosemantic representations of true features.

    Examples:
    >>> input_activations = torch.tensor([[3.0, 4]])
    >>> output_activations = torch.tensor([[1.0, 5]])
    >>> reconstruction_loss(input_activations, output_activations)
    tensor(2.5000)

    Args:
        input_activations: Input activations.
        output_activations: Reconstructed activations.

    Returns:
        Mean Squared Error reconstruction loss.
    """
    return mse_loss(input_activations, output_activations, reduction="mean")


def l1_loss(learned_activations: Float[Tensor, "*batch learned_activations"]) -> Tensor:
    """L1 Loss on Learned Activations.

    L1 loss penalty is the absolute sum of the learned activations. The L1 penalty is this
    multiplied by the l1_coefficient (designed to encourage sparsity).

    Examples:
    >>> learned_activations = torch.tensor([[2.0, -3]])
    >>> l1_loss(learned_activations)
    tensor([5.])

    Args:
        learned_activations: Activations from the hidden layer.

    Returns:
        L1 loss on learned activations.
    """
    return torch.abs(learned_activations).sum(dim=-1)


def sae_training_loss(
    reconstruction_loss_mse: Tensor,
    l1_loss_learned_activations: Tensor,
    l1_coefficient: float,
) -> Tensor:
    """Loss Function for the Sparse Autoencoder.

    The original paper used L2 reconstruction loss, plus l1 loss on the hidden (learned)
    activations.

    https://transformer-circuits.pub/2023/monosemantic-features/index.html#setup-autoencoder-motivation

    Examples:
    >>> reconstruction_loss_mse = torch.tensor([2.5000])
    >>> l1_loss_learned_activations = torch.tensor([1.])
    >>> l1_coefficient = 0.5
    >>> sae_training_loss(reconstruction_loss_mse, l1_loss_learned_activations, l1_coefficient)
    tensor(3.)

    Args:
        reconstruction_loss_mse: MSE reconstruction loss.
        l1_loss_learned_activations: L1 loss on learned activations.
        l1_coefficient: L1 coefficient. The original paper experimented with L1 coefficients of
            [0.01, 0.008, 0.006, 0.004, 0.001]. They used 250 tokens per prompt, so as an
            approximate guide if you use e.g. 2x this number of tokens you might consider using 0.5x
            the l1 coefficient.

    Returns:
        Overall training loss.
    """
    total_loss = reconstruction_loss_mse + l1_loss_learned_activations * l1_coefficient
    return total_loss.sum()
