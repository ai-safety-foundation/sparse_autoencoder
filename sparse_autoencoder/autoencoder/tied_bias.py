"""Tied Biases (Pre-Encoder and Post-Decoder)."""
from jaxtyping import Float
from torch import Tensor
from torch.nn import Module


class PreEncoderBias(Module):
    """Tied Pre-Encoder Bias Layer.

    The tied pre-encoder bias is a learned bias term that is subtracted from the input before
    encoding.

    https://transformer-circuits.pub/2023/monosemantic-features/index.html#appendix-autoencoder-bias

    Args:
        bias: Tied bias parameter (initialised in the parent module), used for both the pre-encoder
            and post-encoder bias. The original paper initialised this using the geometric median of
            the dataset.
    """

    bias: Float[Tensor, " input_activations"]

    def __init__(
        self,
        bias: Float[Tensor, " input_activations"],
    ) -> None:
        """Initialize the bias layer."""
        super().__init__()

        self.bias = bias

    def forward(
        self,
        x: Float[Tensor, "*batch input_activations"],
    ) -> Float[Tensor, "*batch input_activations"]:
        """Forward Pass."""
        return x - self.bias


class PostEncoderBias(Module):
    """Tied Post-Encoder Bias Layer.

    The tied post-encoder bias is a learned bias term that is added to the output of the decoder.

    https://transformer-circuits.pub/2023/monosemantic-features/index.html#appendix-autoencoder-bias

    Args:
        bias: Tied bias parameter (initialised in the parent module), used for both the pre-encoder
            and post-encoder bias. The original paper initialised this using the geometric median of
            the dataset.
    """

    bias: Float[Tensor, " input_activations"]

    def __init__(
        self,
        bias: Float[Tensor, " input_activations"],
    ) -> None:
        """Initialize the bias layer."""
        super().__init__()

        self.bias = bias

    def forward(
        self,
        x: Float[Tensor, "*batch input_activations"],
    ) -> Float[Tensor, "*batch input_activations"]:
        """Forward Pass."""
        return x + self.bias
