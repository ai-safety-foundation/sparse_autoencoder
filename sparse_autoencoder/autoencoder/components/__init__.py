"""Sparse autoencoder components."""
from sparse_autoencoder.autoencoder.components.abstract_decoder import AbstractDecoder
from sparse_autoencoder.autoencoder.components.abstract_encoder import AbstractEncoder
from sparse_autoencoder.autoencoder.components.abstract_outer_bias import AbstractOuterBias
from sparse_autoencoder.autoencoder.components.linear_encoder import LinearEncoder
from sparse_autoencoder.autoencoder.components.tied_bias import TiedBias, TiedBiasPosition
from sparse_autoencoder.autoencoder.components.unit_norm_decoder import UnitNormDecoder


__all__ = [
    "AbstractDecoder",
    "AbstractEncoder",
    "AbstractOuterBias",
    "LinearEncoder",
    "TiedBias",
    "TiedBiasPosition",
    "UnitNormDecoder",
]
