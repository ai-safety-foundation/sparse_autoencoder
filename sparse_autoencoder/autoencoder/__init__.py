"""Sparse autoencoder model & components."""
from sparse_autoencoder.autoencoder.abstract_autoencoder import AbstractAutoencoder
from sparse_autoencoder.autoencoder.components import (
    AbstractDecoder,
    AbstractEncoder,
    AbstractOuterBias,
    LinearEncoder,
    TiedBias,
    TiedBiasPosition,
    UnitNormDecoder,
)
from sparse_autoencoder.autoencoder.model import SparseAutoencoder


__all__ = [
    "AbstractAutoencoder",
    "AbstractDecoder",
    "AbstractEncoder",
    "AbstractOuterBias",
    "SparseAutoencoder",
    "LinearEncoder",
    "TiedBias",
    "TiedBiasPosition",
    "UnitNormDecoder",
]
