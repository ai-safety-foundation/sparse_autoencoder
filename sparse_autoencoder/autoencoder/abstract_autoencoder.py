"""Abstract Sparse Autoencoder Model."""
from abc import ABC, abstractmethod

from torch.nn import Module

from sparse_autoencoder.autoencoder.components.abstract_decoder import AbstractDecoder
from sparse_autoencoder.autoencoder.components.abstract_encoder import AbstractEncoder
from sparse_autoencoder.autoencoder.components.abstract_outer_bias import AbstractOuterBias
from sparse_autoencoder.tensor_types import (
    InputOutputActivationBatch,
    LearnedActivationBatch,
)


class AbstractAutoencoder(Module, ABC):
    """Abstract Sparse Autoencoder Model."""

    @property
    @abstractmethod
    def encoder(self) -> AbstractEncoder:
        """Encoder."""

    @property
    @abstractmethod
    def decoder(self) -> AbstractDecoder:
        """Decoder."""

    @property
    @abstractmethod
    def pre_encoder_bias(self) -> AbstractOuterBias:
        """Pre-encoder bias."""

    @property
    @abstractmethod
    def post_decoder_bias(self) -> AbstractOuterBias:
        """Post-decoder bias."""

    @abstractmethod
    def forward(
        self,
        x: InputOutputActivationBatch,
    ) -> tuple[
        LearnedActivationBatch,
        InputOutputActivationBatch,
    ]:
        """Forward Pass.

        Args:
            x: Input activations (e.g. activations from an MLP layer in a transformer model).

        Returns:
            Tuple of learned activations and decoded activations.
        """

    @abstractmethod
    def reset_parameters(self) -> None:
        """Reset the parameters."""
