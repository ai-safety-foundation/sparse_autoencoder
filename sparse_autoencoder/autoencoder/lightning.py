"""PyTorch Lightning module for training a sparse autoencoder."""
from functools import partial
from typing import Any

from jaxtyping import Float
from lightning.pytorch import LightningModule
from pydantic import NonNegativeFloat, NonNegativeInt, PositiveInt
from torch import Tensor
from torch.optim.optimizer import Optimizer
from torchmetrics import MetricCollection
import wandb

from sparse_autoencoder.activation_resampler.activation_resampler import (
    ActivationResampler,
    ParameterUpdateResults,
)
from sparse_autoencoder.autoencoder.model import (
    ForwardPassResult,
    SparseAutoencoder,
    SparseAutoencoderConfig,
)
from sparse_autoencoder.autoencoder.types import ResetOptimizerParameterDetails
from sparse_autoencoder.metrics.loss.l1_absolute_loss import L1AbsoluteLoss
from sparse_autoencoder.metrics.loss.l2_reconstruction_loss import L2ReconstructionLoss
from sparse_autoencoder.metrics.loss.sae_loss import SparseAutoencoderLoss
from sparse_autoencoder.metrics.train.l0_norm import L0NormMetric
from sparse_autoencoder.metrics.train.neuron_activity import NeuronActivityMetric
from sparse_autoencoder.metrics.wrappers.classwise import ClasswiseWrapperWithMean
from sparse_autoencoder.optimizer.adam_with_reset import AdamWithReset
from sparse_autoencoder.tensor_types import Axis


class LitSparseAutoencoderConfig(SparseAutoencoderConfig):
    """PyTorch Lightning Sparse Autoencoder config."""

    component_names: list[str]

    l1_coefficient: float = 0.001

    resample_interval: PositiveInt = 200000000

    max_n_resamples: NonNegativeInt = 4

    resample_dead_neurons_dataset_size: PositiveInt = 100000000

    resample_loss_dataset_size: PositiveInt = 819200

    resample_threshold_is_dead_portion_fires: NonNegativeFloat = 0.0
    
    normalize_by_input_norm: bool = False

    def model_post_init(self, __context: Any) -> None:  # noqa: ANN401
        """Model post init validation.

        Args:
            __context: Pydantic context.

        Raises:
            ValueError: If the number of component names does not match the number of components.
        """
        if self.n_components and len(self.component_names) != self.n_components:
            error_message = (
                f"Number of component names ({len(self.component_names)}) must match the number of "
                f"components ({self.n_components})"
            )
            raise ValueError(error_message)


class LitSparseAutoencoder(LightningModule):
    """Lightning Sparse Autoencoder."""

    sparse_autoencoder: SparseAutoencoder

    config: LitSparseAutoencoderConfig

    loss_fn: SparseAutoencoderLoss

    train_metrics: MetricCollection

    def __init__(
        self,
        config: LitSparseAutoencoderConfig,
    ):
        """Initialise the module."""
        super().__init__()
        self.sparse_autoencoder = SparseAutoencoder(config)
        self.config = config

        num_components = config.n_components or 1
        add_component_names = partial(
            ClasswiseWrapperWithMean, component_names=config.component_names
        )

        # Create the loss & metrics
        self.loss_fn = SparseAutoencoderLoss(
            num_components, config.l1_coefficient, keep_batch_dim=True, config.normalize_by_input_norm
        )

        self.train_metrics = MetricCollection(
            {
                "l0": add_component_names(L0NormMetric(num_components), prefix="train/l0_norm"),
                "activity": add_component_names(
                    NeuronActivityMetric(config.n_learned_features, num_components),
                    prefix="train/neuron_activity",
                ),
                "l1": add_component_names(
                    L1AbsoluteLoss(num_components), prefix="loss/l1_learned_activations"
                ),
                "l2": add_component_names(
                    L2ReconstructionLoss(num_components), prefix="loss/l2_reconstruction"
                ),
                "loss": add_component_names(
                    SparseAutoencoderLoss(num_components, config.l1_coefficient),
                    prefix="loss/total",
                ),
            },
            # Share state & updates across groups (to avoid e.g. computing l1 twice for both the
            # loss and l1 metrics). Note the metric that goes first must calculate all the states
            # needed by the rest of the group.
            compute_groups=[
                ["loss", "l1", "l2"],
                ["activity"],
                ["l0"],
            ],
        )

        self.activation_resampler = ActivationResampler(
            n_learned_features=config.n_learned_features,
            n_components=num_components,
            resample_interval=config.resample_interval,
            max_n_resamples=config.max_n_resamples,
            n_activations_activity_collate=config.resample_dead_neurons_dataset_size,
            resample_dataset_size=config.resample_loss_dataset_size,
            threshold_is_dead_portion_fires=config.resample_threshold_is_dead_portion_fires,
        )

    def forward(  # type: ignore[override]
        self,
        inputs: Float[
            Tensor, Axis.names(Axis.BATCH, Axis.COMPONENT_OPTIONAL, Axis.INPUT_OUTPUT_FEATURE)
        ],
    ) -> ForwardPassResult:
        """Forward pass."""
        return self.sparse_autoencoder.forward(inputs)

    def update_parameters(self, parameter_updates: list[ParameterUpdateResults]) -> None:
        """Update the parameters of the model from the results of the resampler.

        Args:
            parameter_updates: Parameter updates from the resampler.

        Raises:
            TypeError: If the optimizer is not an AdamWithReset.
        """
        for component_idx, component_parameter_update in enumerate(parameter_updates):
            # Update the weights and biases
            self.sparse_autoencoder.encoder.update_dictionary_vectors(
                component_parameter_update.dead_neuron_indices,
                component_parameter_update.dead_encoder_weight_updates,
                component_idx=component_idx,
            )
            self.sparse_autoencoder.encoder.update_bias(
                component_parameter_update.dead_neuron_indices,
                component_parameter_update.dead_encoder_bias_updates,
                component_idx=component_idx,
            )
            self.sparse_autoencoder.decoder.update_dictionary_vectors(
                component_parameter_update.dead_neuron_indices,
                component_parameter_update.dead_decoder_weight_updates,
                component_idx=component_idx,
            )

            # Reset the optimizer
            for (
                parameter,
                axis,
            ) in self.reset_optimizer_parameter_details:
                optimizer = self.optimizers(use_pl_optimizer=False)
                if not isinstance(optimizer, AdamWithReset):
                    error_message = "Cannot reset the optimizer. "
                    raise TypeError(error_message)

                optimizer.reset_neurons_state(
                    parameter=parameter,
                    neuron_indices=component_parameter_update.dead_neuron_indices,
                    axis=axis,
                    component_idx=component_idx,
                )

    def training_step(  # type: ignore[override]
        self,
        batch: Float[
            Tensor, Axis.names(Axis.BATCH, Axis.COMPONENT_OPTIONAL, Axis.INPUT_OUTPUT_FEATURE)
        ],
        batch_idx: int | None = None,  # noqa: ARG002
    ) -> Float[Tensor, Axis.SINGLE_ITEM]:
        """Training step."""
        # Forward pass
        output: ForwardPassResult = self.forward(batch)

        # Metrics & loss
        train_metrics = self.train_metrics.forward(
            source_activations=batch,
            learned_activations=output.learned_activations,
            decoded_activations=output.decoded_activations,
        )

        loss = self.loss_fn.forward(
            source_activations=batch,
            learned_activations=output.learned_activations,
            decoded_activations=output.decoded_activations,
        )

        if wandb.run is not None:
            self.log_dict(train_metrics)

        # Resample dead neurons
        parameter_updates = self.activation_resampler.forward(
            input_activations=batch,
            learned_activations=output.learned_activations,
            loss=loss,
            encoder_weight_reference=self.sparse_autoencoder.encoder.weight,
        )
        if parameter_updates is not None:
            self.update_parameters(parameter_updates)

        # Return the mean loss
        return loss.mean()

    def on_after_backward(self) -> None:
        """After-backward pass hook."""
        self.sparse_autoencoder.post_backwards_hook()

    def configure_optimizers(self) -> Optimizer:
        """Configure the optimizer."""
        return AdamWithReset(
            self.sparse_autoencoder.parameters(),
            named_parameters=self.sparse_autoencoder.named_parameters(),
            has_components_dim=True,
        )

    @property
    def reset_optimizer_parameter_details(self) -> list[ResetOptimizerParameterDetails]:
        """Reset optimizer parameter details."""
        return self.sparse_autoencoder.reset_optimizer_parameter_details
