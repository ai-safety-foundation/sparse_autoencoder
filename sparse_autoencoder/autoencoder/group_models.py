"""Group of sparse autoencoders."""
from jaxtyping import Float
from torch import Tensor, nn

from sparse_autoencoder.autoencoder.abstract_autoencoder import AbstractAutoencoder
from sparse_autoencoder.tensor_types import Axis


class GroupAutoencoderModels(nn.Module):
    """Group of sparse autoencoders."""

    def __init__(self, models: list[AbstractAutoencoder]):
        """Initialize the group of sparse autoencoders."""
        super().__init__()
        self.models = nn.ModuleList(models)

    def forward(
        self,
        x: list[Float[Tensor, Axis.names(Axis.BATCH, Axis.INPUT_OUTPUT_FEATURE)]],
    ) -> list[
        tuple[
            Float[Tensor, Axis.names(Axis.BATCH, Axis.LEARNT_FEATURE)],
            Float[Tensor, Axis.names(Axis.BATCH, Axis.INPUT_OUTPUT_FEATURE)],
        ]
    ]:
        """Forward Pass.

        Args:
            x: Input activations (e.g. activations from all residual stream layers in a transformer
                model).

        Returns:
            Tuple of learned activations and decoded activations.
        """
        return [model(x) for model in self.models]
