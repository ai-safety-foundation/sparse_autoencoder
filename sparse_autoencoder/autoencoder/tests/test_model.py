"""Sparse Autoencoder Model Tests."""
from syrupy.session import SnapshotSession
import torch

from sparse_autoencoder.autoencoder.model import SparseAutoencoder


def test_initialize_tied_bias() -> None:
    """Check the tied bias is initialised correctly."""
    geometric_median = torch.tensor([1.0, 2.0, 3.0])
    model = SparseAutoencoder(3, 6, geometric_median)
    assert torch.allclose(model.tied_bias, geometric_median)


def test_encoded_decoded_shape_same() -> None:
    """Check the input and output are the same shape."""
    geometric_median = torch.tensor([1.0, 2.0, 3.0])
    model = SparseAutoencoder(3, 6, geometric_median)
    input_tensor = torch.tensor([[2.0, 3.0, 4.0], [5.0, 6.0, 7.0]])
    output = model(input_tensor)

    assert output[1].shape == input_tensor.shape


def test_can_get_encoder_weights() -> None:
    """Check we can access the encoder weights."""
    geometric_median = torch.tensor([1.0, 2.0, 3.0])
    model = SparseAutoencoder(3, 6, geometric_median)
    encoder = model.encoder
    assert encoder.weight.shape == (6, 3)


def test_representation(snapshot: SnapshotSession) -> None:
    """Check the string representation of the model."""
    model = SparseAutoencoder(3, 6, torch.tensor([1.0, 2.0, 3.0]))
    assert snapshot == str(model), "Model string representation has changed."
