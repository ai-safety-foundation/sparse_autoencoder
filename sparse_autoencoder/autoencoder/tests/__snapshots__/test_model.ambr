# serializer version: 1
# name: test_representation
  '''
  SparseAutoencoder(
    (_pre_encoder_bias): TiedBias(position=pre_encoder)
    (_encoder): LinearEncoder(
      in_features=3, out_features=6
      (activation_function): ReLU()
    )
    (_decoder): UnitNormDecoder(in_features=6, out_features=3)
    (_post_decoder_bias): TiedBias(position=post_decoder)
  )
  '''
# ---
